mapper/pchain/tx_parser.go

```diff
@@ -0,0 +1,740 @@
+	if err != nil {
+		return nil, err
+	}
+	addMetadataToStakeOuts(ops, &tx.Validator)
+
+	return ops, nil
+}
+
+func (t *TxParser) parseRewardValidatorTx(tx *txs.RewardValidatorTx) (*txOps, error) {
+	stakingTxID := tx.TxID
+
+	if t.dependencyTxs == nil {
+		return nil, errNoDependencyTxs
+	}
+	rewardOuts := t.dependencyTxs[stakingTxID.String()]
+	if rewardOuts == nil {
+		return nil, errNoMatchingRewardOutputs
+	}
+	ops := newTxOps(t.isConstruction)
+	err := t.utxosToOperations(ops, OpRewardValidator, rewardOuts.RewardUTXOs, OpTypeReward, mapper.PChainNetworkIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	var validator *pChainValidator.Validator
+	switch utx := rewardOuts.Tx.Unsigned.(type) {
+	case *txs.AddValidatorTx:
+		validator = &utx.Validator
+	case *txs.AddDelegatorTx:
+		validator = &utx.Validator
+	}
+	addMetadataToStakeOuts(ops, validator)
+
+	return ops, nil
+}
+
+func addMetadataToStakeOuts(ops *txOps, validator *pChainValidator.Validator) {
+	if validator != nil {
+		for _, out := range ops.StakeOuts {
+			out.Metadata[MetadataValidatorNodeID] = validator.NodeID.String()
+			out.Metadata[MetadataStakingStartTime] = validator.Start
+			out.Metadata[MetadataStakingEndTime] = validator.End
+		}
+	}
+}
+
+func (t *TxParser) parseCreateSubnetTx(txID ids.ID, tx *txs.CreateSubnetTx) (*txOps, error) {
+	return t.baseTxToCombinedOperations(txID, &tx.BaseTx, OpCreateSubnet)
+}
+
+func (t *TxParser) parseAddSubnetValidatorTx(txID ids.ID, tx *txs.AddSubnetValidatorTx) (*txOps, error) {
+	return t.baseTxToCombinedOperations(txID, &tx.BaseTx, OpAddSubnetValidator)
+}
+
+func (t *TxParser) parseRemoveSubnetValidatorTx(txID ids.ID, tx *txs.RemoveSubnetValidatorTx) (*txOps, error) {
+	return t.baseTxToCombinedOperations(txID, &tx.BaseTx, OpRemoveSubnetValidator)
+}
+
+func (t *TxParser) parseTransformSubnetTx(txID ids.ID, tx *txs.TransformSubnetTx) (*txOps, error) {
+	return t.baseTxToCombinedOperations(txID, &tx.BaseTx, OpTransformSubnetValidator)
+}
+
+func (t *TxParser) parseCreateChainTx(txID ids.ID, tx *txs.CreateChainTx) (*txOps, error) {
+	return t.baseTxToCombinedOperations(txID, &tx.BaseTx, OpCreateChain)
+}
+
+func (t *TxParser) baseTxToCombinedOperations(txID ids.ID, tx *txs.BaseTx, txType string) (*txOps, error) {
+	ops := newTxOps(t.isConstruction)
+
+	err := t.insToOperations(ops, txType, tx.Ins, OpTypeInput)
+	if err != nil {
+		return nil, err
+	}
+
+	err = t.outsToOperations(ops, txType, txID, tx.Outs, OpTypeOutput, mapper.PChainNetworkIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	return ops, nil
+}
+
+func (t *TxParser) insToOperations(
+	inOps *txOps,
+	opType string,
+	txIns []*avax.TransferableInput,
+	metaType string,
+) error {
+	status := types.String(mapper.StatusSuccess)
+	if t.isConstruction {
+		status = nil
+	}
+
+	for _, in := range txIns {
+		metadata := &OperationMetadata{
+			Type: metaType,
+		}
+
+		if transferInput, ok := in.In.(*secp256k1fx.TransferInput); ok {
```

Comment:
What if this input is a `stakeable.LockIn` and not a `*secp256k1fx.TransferInput`?

Reply:
Indeed I have opened the PR https://github.com/ava-labs/avalanche-rosetta/pull/168 to fix this. @bilgehansahin please have a look